examples/LeNet/LeNet.cpp

```diff
@@ -0,0 +1,180 @@
+// Copyright 2021 The WebNN-native Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "examples/LeNet/LeNet.h"
+
+#include <stdlib.h>
+#include <chrono>
+
+#include "common/Log.h"
+
+const size_t WEIGHTS_LENGTH = 1724336;
+
+LeNet::LeNet() {
+    mContext = CreateCppNeuralNetworkContext();
+    mContext.SetUncapturedErrorCallback(
+        [](WebnnErrorType type, char const* message, void* userData) {
+            if (type != WebnnErrorType_NoError) {
+                dawn::ErrorLog() << "Error type is " << type << ", message is " << message;
+            }
+        },
+        this);
+}
+
+bool LeNet::Load(const std::string& weigthsPath) {
```

Comment:
Couple questions:
1. Should WebNN consider defining a model format that covers both topology and weights? This should simplify model loading.
2. There seems to missing TensorShape class that provide shape operations like size, rank, ...
3. For async calls like `compile` `compute`, any idea how the callback would work in the WASM env?

Reply:
It was explicitly stated as a non-goal for webnn to directly support model format as we believe formats should be left with the framework or the app using it. I'll let @huningxin answer the other questions.

Reply:
> 2\. There seems to missing TensorShape class that provide shape operations like size, rank, ...

This would leave to user code to calculate the tensor size and rank. As `webnn.h` aims at an 1-to-1 mapping to WebNN spec, user code could use [`OperandDescriptor.dimensions`](https://webmachinelearning.github.io/webnn/#dictdef-operanddescriptor) to calculate the size and rank. As [the LeNet sample](https://github.com/webmachinelearning/webnn-native/blob/beb8ad99af24f11b33c6306a903c640d647d75ca/examples/LeNet/LeNet.cpp#L56) shows, the shape of a conv2d filter could be defined as:
```c++
const std::vector<int32_t> conv2d1FilterShape = {20, 1, 5, 5};
```
Then its rank would be:
```c++
size_t rank = conv2d1FilterShape.size();
```
The number of elements would be:
```c++
size_t elements = std::accumulate(conv2d1FilterShape.begin(), conv2d1FilterShape.end(), size_t(1), std::multiplies<size_t>());
```
As its type is `float32`, its size would be:
```c++
size_t size = elements * sizeof(float);
```

Reply:
> 3\. For async calls like `compile` `compute`, any idea how the callback would work in the WASM env?

This would be similar to WebGPU async calls in the Wasm env.  The callback would be invoked by the `.then` handler of the `promise` returned by `compile` and `compute`. For example, WebGPU `mapAsync` implements that in the [`library_webgpu.js`](https://github.com/emscripten-core/emscripten/blob/main/src/library_webgpu.js#L1483) of emscripten. We will add the `library_webnn.js` into emscripten later.